v5.11.3: 发布超时后音视频 writer/frame 空指针可导致进程崩溃

## 环境

- `github.com/cuteLittleDevil/m7s-jt1078/v5 v5.11.3`（当前 `main@e0c7da5` 仍受影响）
- `m7s.live/v5 v5.1.3`
- Linux/amd64
- 同一终端 6 路 JT1078 实时视频并发

## 现象

多路连接建立后，网关先记录 `unqualified data`、`video timeout`、`audio timeout`、`no IDRring`，随后一个晚到 H.264/H.265 包触发 panic，整个 Go 进程退出：

```text
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x50]

goroutine 164 [running]:
github.com/cuteLittleDevil/m7s-jt1078/v5/pkg.(*connection).handle(...)
    github.com/cuteLittleDevil/m7s-jt1078/v5@v5.11.3/pkg/connection.go:193 +0x26a
github.com/cuteLittleDevil/m7s-jt1078/v5/pkg.(*connection).run-range1(...)
    github.com/cuteLittleDevil/m7s-jt1078/v5@v5.11.3/pkg/connection.go:104
```

## 根因

对同版本 Linux/amd64 二进制反汇编后，`handle+0x26a` 精确对应 `frame.Timestamp` 的 `MOVQ 0x50(CX), BX`。故障地址也是 `0x50`，可确认 nil 对象是 `writer.VideoFrame`。

`m7s.live/v5` 在 `video timeout` 后通过 `Publisher.NoVideo()` 将 `PubVideo=false`。`PublishVideoWriter.NextVideo()` 随后调用 `getVideoFrameToWrite()`，该函数在 `!PubVideo` 时返回 nil；下一个晚到视频包在 `connection.go:193` 无保护解引用该 nil frame。

音频路径的 `writer.AudioFrame` 使用相同模式，也存在 writer/frame 无保护解引用。除此之外，`Service.Run` 启动的每连接 goroutine 没有 panic 隔离，任一连接 panic 会终止整个网关进程。

## 期望行为

- publisher、audio/video writer 或 frame 不可用时返回明确错误并结束该连接，不发生 panic；
- 单连接出现未预期 panic 时记录 panic 与完整 stack，其他连接及主进程继续运行。

## 建议修复与测试

1. 在音频和视频写入前分别校验 publisher、writer、frame。
2. 在每连接 goroutine 边界增加 `recover` 与 `debug.Stack()` 日志。
3. 覆盖 publisher 不可用、writer 为 nil、frame 为 nil，以及连接 panic 隔离的回归测试。

我已准备最小修复和上述测试，随后提交 PR。

--- a/v5/pkg/connection.go
+++ b/v5/pkg/connection.go
@@ -18,6 +18,14 @@ import (
 	"time"
 )
 
+var (
+	errPublisherUnavailable   = errors.New("publisher unavailable")
+	errAudioWriterUnavailable = errors.New("audio writer unavailable")
+	errAudioFrameUnavailable  = errors.New("audio frame unavailable")
+	errVideoWriterUnavailable = errors.New("video writer unavailable")
+	errVideoFrameUnavailable  = errors.New("video frame unavailable")
+)
+
 type connection struct {
 	conn net.Conn
 	*slog.Logger
@@ -161,16 +169,49 @@ func (c *connection) stop() {
 	})
 }
 
+func (c *connection) getAudioWriter() (*m7s.PublishAudioWriter[*format.Mpeg2Audio], error) {
+	if c.publisher == nil {
+		return nil, errPublisherUnavailable
+	}
+	c.audioWriterOnce.Do(func() {
+		allocator := gomem.NewScalableMemoryAllocator(1 << gomem.MinPowerOf2)
+		c.audioWriter = m7s.NewPublishAudioWriter[*format.Mpeg2Audio](c.publisher, allocator)
+	})
+	if c.audioWriter == nil {
+		return nil, errAudioWriterUnavailable
+	}
+	if c.audioWriter.AudioFrame == nil {
+		return nil, errAudioFrameUnavailable
+	}
+	return c.audioWriter, nil
+}
+
+func (c *connection) getVideoWriter() (*m7s.PublishVideoWriter[*format.AnnexB], error) {
+	if c.publisher == nil {
+		return nil, errPublisherUnavailable
+	}
+	c.videoWriterOnce.Do(func() {
+		allocator := gomem.NewScalableMemoryAllocator(1 << gomem.MinPowerOf2)
+		c.videoWriter = m7s.NewPublishVideoWriter[*format.AnnexB](c.publisher, allocator)
+	})
+	if c.videoWriter == nil {
+		return nil, errVideoWriterUnavailable
+	}
+	if c.videoWriter.VideoFrame == nil {
+		return nil, errVideoFrameUnavailable
+	}
+	return c.videoWriter, nil
+}
+
 func (c *connection) handle(packet *jt1078.Packet) error {
 	data := packet.Body
 
 	switch pt := packet.Flag.PT; pt {
 	case jt1078.PTAAC, jt1078.PTG711A, jt1078.PTG711U:
-		c.audioWriterOnce.Do(func() {
-			allocator := gomem.NewScalableMemoryAllocator(1 << gomem.MinPowerOf2)
-			c.audioWriter = m7s.NewPublishAudioWriter[*format.Mpeg2Audio](c.publisher, allocator)
-		})
-		writer := c.audioWriter
+		writer, err := c.getAudioWriter()
+		if err != nil {
+			return err
+		}
 		frame := writer.AudioFrame
 		frame.ICodecCtx = &codec.AACCtx{}
 		if pt == jt1078.PTG711A {
@@ -184,11 +225,10 @@ func (c *connection) handle(packet *jt1078.Packet) error {
 		return writer.NextAudio()
 
 	case jt1078.PTH264, jt1078.PTH265:
-		c.videoWriterOnce.Do(func() {
-			allocator := gomem.NewScalableMemoryAllocator(1 << gomem.MinPowerOf2)
-			c.videoWriter = m7s.NewPublishVideoWriter[*format.AnnexB](c.publisher, allocator)
-		})
-		writer := c.videoWriter
+		writer, err := c.getVideoWriter()
+		if err != nil {
+			return err
+		}
 		frame := writer.VideoFrame
 		frame.Timestamp = c.timestampFunc(packet)
 		// Push raw AnnexB data — do NOT use GetNalus()/PushOne() as that sets

--- a/v5/pkg/connection_test.go
+++ b/v5/pkg/connection_test.go
@@ -1,0 +1,128 @@
+package pkg
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"log/slog"
+	"net"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/cuteLittleDevil/go-jt808/protocol/jt1078"
+	"m7s.live/v5"
+	"m7s.live/v5/pkg/format"
+)
+
+func newTestConnection() *connection {
+	logger := slog.New(slog.NewTextHandler(&bytes.Buffer{}, nil))
+	return newConnection(nil, logger, func(*jt1078.Packet) time.Duration { return 0 })
+}
+
+func TestConnectionHandleRejectsUnavailablePublisher(t *testing.T) {
+	c := newTestConnection()
+	err := c.handle(&jt1078.Packet{Flag: jt1078.Flag{PT: jt1078.PTH264}})
+	if !errors.Is(err, errPublisherUnavailable) {
+		t.Fatalf("handle() error = %v, want %v", err, errPublisherUnavailable)
+	}
+}
+
+func TestConnectionHandleRejectsUnavailableWriters(t *testing.T) {
+	tests := []struct {
+		name string
+		pt   jt1078.PTType
+		want error
+	}{
+		{name: "audio", pt: jt1078.PTAAC, want: errAudioWriterUnavailable},
+		{name: "video", pt: jt1078.PTH264, want: errVideoWriterUnavailable},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			c := newTestConnection()
+			c.publisher = &m7s.Publisher{}
+			err := c.handle(&jt1078.Packet{Flag: jt1078.Flag{PT: tt.pt}})
+			if !errors.Is(err, tt.want) {
+				t.Fatalf("handle() error = %v, want %v", err, tt.want)
+			}
+		})
+	}
+}
+
+func TestConnectionHandleRejectsUnavailableFrames(t *testing.T) {
+	tests := []struct {
+		name    string
+		pt      jt1078.PTType
+		prepare func(*connection)
+		want    error
+	}{
+		{
+			name: "audio",
+			pt:   jt1078.PTAAC,
+			prepare: func(c *connection) {
+				c.publisher.PubAudio = true
+				c.audioWriter = &m7s.PublishAudioWriter[*format.Mpeg2Audio]{Publisher: c.publisher}
+				c.audioWriterOnce.Do(func() {})
+			},
+			want: errAudioFrameUnavailable,
+		},
+		{
+			name: "video",
+			pt:   jt1078.PTH264,
+			prepare: func(c *connection) {
+				c.publisher.PubVideo = true
+				c.videoWriter = &m7s.PublishVideoWriter[*format.AnnexB]{Publisher: c.publisher}
+				c.videoWriterOnce.Do(func() {})
+			},
+			want: errVideoFrameUnavailable,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			c := newTestConnection()
+			c.publisher = &m7s.Publisher{}
+			tt.prepare(c)
+			err := c.handle(&jt1078.Packet{Flag: jt1078.Flag{PT: tt.pt}})
+			if !errors.Is(err, tt.want) {
+				t.Fatalf("handle() error = %v, want %v", err, tt.want)
+			}
+		})
+	}
+}
+
+type panicReadConn struct {
+	net.Conn
+}
+
+func (panicReadConn) Read([]byte) (int, error) {
+	panic("test connection panic")
+}
+
+func (panicReadConn) Close() error {
+	return nil
+}
+
+func TestServiceRunConnectionRecoversPanic(t *testing.T) {
+	var output bytes.Buffer
+	logger := slog.New(slog.NewJSONHandler(&output, nil))
+	service := &Service{Logger: logger}
+	client := newConnection(panicReadConn{}, logger, func(*jt1078.Packet) time.Duration { return 0 })
+	client.onLeaveEvent = func() {}
+	httpBody := map[string]any{"streamPath": "live/test"}
+
+	service.runConnection(context.Background(), client, 0, &httpBody)
+
+	select {
+	case <-client.stopChan:
+	default:
+		t.Fatal("connection was not stopped after panic")
+	}
+	logOutput := output.String()
+	for _, want := range []string{"connection panic", "test connection panic", "runtime/debug.Stack", "live/test"} {
+		if !strings.Contains(logOutput, want) {
+			t.Fatalf("panic log %q does not contain %q", logOutput, want)
+		}
+	}
+}

--- a/v5/pkg/service.go
+++ b/v5/pkg/service.go
@@ -8,6 +8,7 @@ import (
 	"net"
 	"os"
 	"path/filepath"
+	"runtime/debug"
 	"strings"
 	"time"
 )
@@ -33,6 +34,22 @@ type Service struct {
 	*slog.Logger
 	addr string
 	opts *Options
+}
+
+func (s *Service) runConnection(ctx context.Context, client *connection, waitSubscriberOverTime time.Duration, httpBody *map[string]any) {
+	defer func() {
+		if panicValue := recover(); panicValue != nil {
+			s.Error("connection panic",
+				slog.Any("http body", *httpBody),
+				slog.Any("panic", panicValue),
+				slog.String("stack", string(debug.Stack())))
+		}
+	}()
+	if err := client.run(ctx, waitSubscriberOverTime); err != nil {
+		s.Warn("run error",
+			slog.Any("http body", *httpBody),
+			slog.String("err", err.Error()))
+	}
 }
 
 func (s *Service) Run() {
@@ -99,12 +116,6 @@ func (s *Service) Run() {
 			}
 			cancel()
 		}
-		go func(ctx context.Context, waitSubscriberOverTime time.Duration) {
-			if err := client.run(ctx, waitSubscriberOverTime); err != nil {
-				s.Warn("run error",
-					slog.Any("http body", httpBody),
-					slog.String("err", err.Error()))
-			}
-		}(ctx, s.opts.overTime)
+		go s.runConnection(ctx, client, s.opts.overTime, &httpBody)
 	}
 }

